native/src/main/java/io/ballerina/lib/avro/visitor/DeserializeVisitor.java

```diff
@@ -0,0 +1,411 @@
+                                       ((Number) fieldData).longValue());
+                    }
+                }
+                case FLOAT, DOUBLE -> {
+                    if (fieldData instanceof Double) {
+                        avroRecord.put(StringUtils.fromString(field.name()), fieldData);
+                    } else {
+                        avroRecord.put(StringUtils.fromString(field.name()), Double.parseDouble(fieldData.toString()));
+                    }
+                }
+                case ENUM -> {
+                    if (fieldData instanceof GenericEnumSymbol<?>) {
+                        avroRecord.put(StringUtils.fromString(field.name()),
+                                       StringUtils.fromString(fieldData.toString()));
+                    }
+                }
+                default -> {
+                    if (fieldData instanceof Boolean) {
+                        avroRecord.put(StringUtils.fromString(field.name()), fieldData);
+                    }
+                }
+            }
+        }
+    }
+
+    private static Type extractMapType(Type type) {
+        Type mapType = type;
+        for (Map.Entry<String, Field> entry : ((RecordType) type).getFields().entrySet()) {
+            Field fieldValue = entry.getValue();
+            if (fieldValue != null) {
+                Type fieldType = fieldValue.getFieldType();
+                if (fieldType instanceof MapType) {
+                    mapType = fieldType;
+                } else if (TypeUtils.getReferredType(fieldType) instanceof MapType) {
+                    mapType = TypeUtils.getReferredType(fieldType);
+                } else if (fieldType instanceof IntersectionType) {
+                    Type referredType = getMutableType((IntersectionType) fieldType);
+                    if (referredType instanceof MapType) {
+                        mapType = referredType;
+                    }
+                }
+            }
+        }
+        return mapType;
+    }
+
+    private static RecordType extractRecordType(RecordType type) {
+        Map<String, Field> fieldsMap = type.getFields();
+        RecordType recType = type;
+        for (Map.Entry<String, Field> entry : fieldsMap.entrySet()) {
+            Field fieldValue = entry.getValue();
+            if (fieldValue != null) {
+                Type fieldType = fieldValue.getFieldType();
+                if (fieldType instanceof RecordType) {
+                    recType = (RecordType) fieldType;
+                } else if (fieldType instanceof IntersectionType) {
+                    Type getType = getMutableType((IntersectionType) fieldType);
+                    if (getType instanceof RecordType) {
+                        recType = (RecordType) getType;
+                    }
+                } else if (TypeUtils.getReferredType(fieldType) instanceof RecordType) {
+                    recType = (RecordType) TypeUtils.getReferredType(fieldType);
+                }
```

Comment:
Let's use the `typeTag`s instead of `instanceof` checks. (`instanceof` checks are expensive compared to the type tag checks. 
Let's also try to reduce the if-else ladders here.